### PR TITLE
Resolve #67 gpt2 example polish

### LINKED ISSUE #67: gpt2 example polish
- [x] Port [gpt2_train_main.py](https://github.com/asyml/texar/blob/master/examples/gpt-2/gpt2_train_main.py) too

- [x] https://github.com/asyml/texar-pytorch/blob/dc51de295da9e861d739bf40ee4902cf8669ca53/examples/gpt-2/gpt2_generate_main.py#L26-L28
do not expose internal details. ---> `from texar.modules import TransformerDecoder`

- [x] https://github.com/asyml/texar-pytorch/blob/dc51de295da9e861d739bf40ee4902cf8669ca53/examples/gpt-2/gpt2_generate_main.py#L84-L95
Too sparse. Reformat to sth like 
```
parser.add_argument(
    '--config_model',
    ...
```

- [x] https://github.com/asyml/texar-pytorch/blob/dc51de295da9e861d739bf40ee4902cf8669ca53/examples/gpt-2/gpt2_generate_main.py#L100
Should this inherit `nn.Module` or `texar.ModuleBase`?

- [x] https://github.com/asyml/texar-pytorch/blob/dc51de295da9e861d739bf40ee4902cf8669ca53/examples/gpt-2/gpt2_generate_main.py#L197-L198
Better to print a message

### PR DESCRIPTION
**Pull request by gpengzhi**
Add `GPT2` training example.
Fix `doc` issue raised in #110

### CODE REVIEW

File: examples/gpt-2/README.md [hunk 1]
```diff
@@ -122,7 +136,56 @@ in this way".
...
+
+### Prepare data
+
+We first preprocess data with the GPT-2 BPE encoding. 
+
+A toy dataset is provided under [`data/toy/`](data/toy) which includes `train.txt`, `dev.txt`, and `test.txt`. This example will fit the GPT-2 model on `train.txt`, evaluate perplexity on `dev.txt`, and do continuation generation using `test.txt` as the context.
+
+Run the following cmd to transform the data into [TFRecord](https://www.tensorflow.org/tutorials/load_data/tf_records) format and perform processing such as truncation, BPE encoding, adding special tokens, etc:
+
+```
```
Comment (huzecong) [inline examples/gpt-2/README.md]: Use ```` ```bash ```` to add syntax highlighting in GFM.

Also, multiline bash scripts require backslashes (`\`) at end of line.

Finally, please remove the unnecessary indent.

File: examples/gpt-2/README.md [hunk 2]
```diff
@@ -122,7 +136,56 @@ in this way".
...
+    [--tfrecord_output_dir=data/toy] 
+    [--pretrained_model_name=117M]
+```
+
+- `data_dir`: The directory of raw data, wherein data files must be named as 'train.txt', 'dev.txt', or 'test.txt'. It is *not* necessary to provide all three files.
+- `max_seq_length`: The maxium length of sequence after BPE encoding. This includes GPT-2 special tokens that will be automatically added. Longer sequence will be trimmed. 
+- `tfrecord_output_dir`: The output path where the resulting TFRecord files will be put in. Be default, it is set to be the same as `data_dir`. 
+- `pretrained_model_name`: The name of a pre-trained model to load selected in the list of: `117M`, `345M`.
+
+The above cmd will output TFRecord files in the specified output directory. E.g., if `train.txt` is provided under `data_dir`, the output file `train.tf_record` will be produced under `tfrecord_output_dir`.
```
Comment (huzecong) [inline examples/gpt-2/README.md]: We no longer use `TFRecord` files. Now outputs are saved as pickled files I think?

Comment (gpengzhi) [reply examples/gpt-2/README.md]: We might need to modify the `doc` in `RecordData`. The description is outdated. @huzecong

File: examples/gpt-2/README.md [hunk 3]
```diff
@@ -8,9 +8,23 @@ In sum, this example showcases:
 
 * Contructing and using pre-trained GPT-2 models in Texar
 * Using GPT-2 to generate text samples with or without context
-* **Train or fine-tune** the model with **distributed GPU** (coming soon)
+* **Train or fine-tune** the model with **single GPU**
```
Comment (ZhitingHu) [inline examples/gpt-2/README.md]: let's delete `with **single GPU**`

File: examples/gpt-2/README.md [hunk 4]
```diff
@@ -122,7 +139,56 @@ in this way".
...
+We first preprocess data with the GPT-2 BPE encoding. 
+
+A toy dataset is provided under [`data/toy/`](data/toy) which includes `train.txt`, `dev.txt`, and `test.txt`. This example will fit the GPT-2 model on `train.txt`, evaluate perplexity on `dev.txt`, and do continuation generation using `test.txt` as the context.
+
+Run the following cmd to transform the data into [pickle](https://docs.python.org/3/library/pickle.html) format and perform processing such as truncation, BPE encoding, adding special tokens, etc:
+
+```bash
+python prepare_data.py --data_dir data/toy \
+--max_seq_length=128 \
+--tfrecord_output_dir=data/toy \
```
Comment (ZhitingHu) [inline examples/gpt-2/README.md]: `tfrecord_output_dir` -- it shouldn't be `tfrecord`.

Can we just name it `output_dir`? The TF  example could be updated in the same way to make the interfaces the same

File: examples/gpt-2/prepare_data.py [hunk 5]
```diff
@@ -0,0 +1,80 @@
...
+
+
+def prepare_data():
+    r"""Preprocesses raw data and produces TFRecord files.
+    """
+    data_dir = args.data_dir
+    if args.tfrecord_output_dir is None:
+        tfrecord_output_dir = data_dir
+    else:
+        tfrecord_output_dir = args.tfrecord_output_dir
```
Comment (ZhitingHu) [inline examples/gpt-2/prepare_data.py]: Let's rename these `tfrecord_xx` cuz they are not `tfrecord`

File: examples/bert/README.md [hunk 6]
```diff
@@ -10,6 +10,16 @@ In sum, this example showcases:
 * Building and fine-tuning on downstream tasks
 * Use of Texar `TFRecordData` module for data loading and processing
```
Comment (ZhitingHu) [inline examples/bert/README.md]: No `TFRecord*`

File: examples/gpt-2/README.md [hunk 7]
```diff
@@ -122,7 +139,56 @@ in this way".
...
+--output_dir=data/toy \ 
+--pretrained_model_name=117M
+```
+
+- `data_dir`: The directory of raw data, wherein data files must be named as 'train.txt', 'dev.txt', or 'test.txt'. It is *not* necessary to provide all three files.
+- `max_seq_length`: The maxium length of sequence after BPE encoding. This includes GPT-2 special tokens that will be automatically added. Longer sequence will be trimmed. 
+- `output_dir`: The output path where the resulting pickle files will be put in. Be default, it is set to be the same as `data_dir`. 
+- `pretrained_model_name`: The name of a pre-trained model to load selected in the list of: `117M`, `345M`.
+
+The above cmd will output pickle files in the specified output directory. E.g., if `train.txt` is provided under `data_dir`, the output file `train.pickle` will be produced under `tfrecord_output_dir`.
```
Comment (ZhitingHu) [inline examples/gpt-2/README.md]: no `tfrecord*`

File: examples/gpt-2/README.md [hunk 8]
```diff
@@ -176,12 +242,10 @@ output, output_length = decoder(
 **Ex. Use 3): Fine-tuning for conditional generation**
 
 ```python
-tgt_embed = decoder.word_embedder(truth_target[:, :-1]) + decoder.position_embedder(sequence_length=tgt_len-1)
-
 output = decoder(
     memory=source_hidden_states, 
     memory_sequence_length=src_len,
-    inputs=tgt_embed,
+    inputs=input_ids,
```
Comment (ZhitingHu) [inline examples/gpt-2/README.md]: Is this correct? Where does the decoder get access to embedders if the inputs are just `input_ids`?

Comment (gpengzhi) [reply examples/gpt-2/README.md]: https://github.com/asyml/texar-pytorch/blob/master/texar/modules/decoders/gpt2_decoder.py#L438

Comment (gpengzhi) [reply examples/gpt-2/README.md]: Embedders are in the module.

File: examples/bert/README.md (hunk 6 above)
Comment (gpengzhi) [reply examples/bert/README.md]: For `bert` example, I believe there are more TFrecord related issues. I am currently refactoring the `pretrain` module and will polish this example later.

**Pull request merged**